lib/integrations/awsra/profile_syncer.go

```diff
@@ -0,0 +1,418 @@
+		select {
+		case <-ctx.Done():
+			return nil
+
+		case <-params.Clock.After(params.SyncPollInterval):
+		}
+	}
+}
+
+func fetchProxyPublicAddr(cache SyncerCache) (string, error) {
+	proxies, err := cache.GetProxies()
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	if len(proxies) == 0 {
+		return "", trace.NotFound("no proxies found")
+	}
+
+	proxy := proxies[0]
+	if proxy.GetPublicAddr() == "" {
+		return "", trace.NotFound("proxy %q does not have a public address", proxy.GetName())
+	}
+	return proxy.GetPublicAddr(), nil
+}
+
+func integrationsWithProfileSyncEnabled(ctx context.Context, cache SyncerCache) ([]types.Integration, error) {
+	var integrations []types.Integration
+	var nextKey string
+
+	for {
+		resp, respNextKey, err := cache.ListIntegrations(ctx, 0, nextKey)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		for _, integration := range resp {
+			if integration.GetSubKind() != types.IntegrationSubKindAWSRolesAnywhere ||
+				integration.GetAWSRolesAnywhereIntegrationSpec().ProfileSyncConfig == nil ||
+				!integration.GetAWSRolesAnywhereIntegrationSpec().ProfileSyncConfig.Enabled {
+
+				continue
+			}
+
+			integrations = append(integrations, integration)
+		}
+
+		if respNextKey == "" {
+			break
+		}
+		nextKey = respNextKey
+	}
+
+	return integrations, nil
+}
+
+func buildAWSRolesAnywhereClientForIntegration(ctx context.Context, params AWSRolesAnywherProfileSyncerParams, integration types.Integration) (RolesAnywhereClient, error) {
+	trustAnchorARN := integration.GetAWSRolesAnywhereIntegrationSpec().TrustAnchorARN
+	profileSyncProfileARN := integration.GetAWSRolesAnywhereIntegrationSpec().ProfileSyncConfig.ProfileARN
+	profileSyncRoleARN := integration.GetAWSRolesAnywhereIntegrationSpec().ProfileSyncConfig.RoleARN
+	profileAcceptsRoleSessionName := integration.GetAWSRolesAnywhereIntegrationSpec().ProfileSyncConfig.ProfileAcceptsRoleSessionName
+
+	parsedProfileSyncProfile, err := arn.Parse(profileSyncProfileARN)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	region := parsedProfileSyncProfile.Region
+
+	resp, err := GenerateCredentials(ctx, GenerateCredentialsRequest{
+		Clock:                 params.Clock,
+		TrustAnchorARN:        trustAnchorARN,
+		ProfileARN:            profileSyncProfileARN,
+		RoleARN:               profileSyncRoleARN,
+		SubjectCommonName:     params.subjectName,
+		AcceptRoleSessionName: profileAcceptsRoleSessionName,
+		KeyStoreManager:       params.KeyStoreManager,
+		Cache:                 params.Cache,
+		CreateSession:         params.createSession,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	awsConfig, err := config.LoadDefaultConfig(
+		ctx,
+		config.WithRegion(region),
+		config.WithCredentialsProvider(credentials.NewStaticCredentialsProvider(resp.AccessKeyID, resp.SecretAccessKey, resp.SessionToken)),
+	)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if params.rolesAnywhereClient != nil {
+		return params.rolesAnywhereClient, nil
+	}
```

Comment:
Can this be earlier in the function, since it doesn't rely on the generated credentials?

Reply:
This is only used for testing and I was trying to delay as much as possible its usage to ensure we test everything that comes before this section.

Reply:
👍 I'd recommend adding a comment then, so future readers will understand why it's here.